R/palaeorotate.R

```diff
@@ -318,22 +320,32 @@ palaeorotate <- function(occdf, lng = "lng", lat = "lat", age = "age",
         which.min(abs(occdf[i, age] - rot_age))
       })]
 
-    # Search for matching longitude and  latitude
-    occdf$rot_lng <- sapply(seq_len(nrow(occdf)), function(i) {
-      # Extract closest longitude
-      base_model[which.min(abs(base_model[, c("lng")]  - occdf[i, lng])), 1]
-    }, simplify = TRUE)
+    # Set-up
+    # Convert to sf object and add CRS
+    occdf_sf <- sf::st_as_sf(x = occdf,
+                             coords = c(lng, lat),
+                             remove = FALSE,
+                             crs = "EPSG:4326")
+    base_model_sf <- sf::st_as_sf(x = base_model,
+                                  coords = c("lng", "lat"),
+                                  remove = FALSE,
+                                  crs = "EPSG:4326")
 
-    occdf$rot_lat <- sapply(seq_len(nrow(occdf)), function(i) {
-      # Extract closest latitude
-      base_model[which.min(abs(base_model[, c("lat")]  - occdf[i, lat])), 2]
-    }, simplify = TRUE)
+    # Match points with cells
+    occ_cell <- h3jsr::point_to_cell(input = occdf_sf,
+                                     res = 3, # Grid resolution
+                                     simple = TRUE)
+    model_cell <- h3jsr::point_to_cell(input = base_model_sf,
+                                       res = 3, # Grid resolution
+                                       simple = TRUE)
 
     # Generate row index
     pc_ind <- sapply(seq_len(nrow(occdf)), function(i) {
-      which(base_model[, c("lng")] == occdf[i, "rot_lng"] &
-              base_model[, c("lat")] == occdf[i, "rot_lat"])
+      which(model_cell == occ_cell[i])
```

Comment:
I'm not sure if it's faster, but I think you could use `match` here instead of `sapply` and `which`?

Reply:
Just checked, and it should be orders of magnitude faster.

Reply:
Another idea to speed things up would be to include the cell indices in the data files, so you don't need to do point_to_cell each time you use the function?

Reply:
I've updated to use `match`, and it does seem to be a lot faster! Thanks! 

I agree that including the cell indices in the data files will likely speed things up. However, this would then require depositing another version of the data on Zenodo which I am a bit reluctant to do right now. Considering this function isn't really slowing things down a great deal, I'd prefer to keep it as it is. The addition of the cell indices in the data files is also likely to make them a bit heavier (increasing download time), so it probably doesn't make too much of a difference.

Reply:
Nice! (I updated the benchmarks above)

With regards to the cell indices, I agree it's fine to leave it as-is!